### PR TITLE
Upgrade Postgres to 10 on Red Hat systems

### PR DESCRIPTION
**Pull request by ehelms**
Just a starting place for the upgrade workflow and approach. The upgrade this way *must* happen after the migrate Candlepin and Foreman steps which is a bit ugly. Changing this requires me to move these into being handled by the puppet modules / installation steps.

### CODE REVIEW

**Review by ekohl — COMMENTED** [review]
If we put it in hiera this way, we also need to upgrade it on vanilla Foreman.

We also need to check if the database is local which means checking the `manage_database` parameter on the foreman class and katello's candlepin. For foreman you also need to make sure it's only ran on EL 7.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb:101]: Can you move this to line 95 so we perform the DB migration before other migrations?

Comment (ehelms) [reply katello/hooks/pre/30-upgrade.rb:101]: See https://github.com/theforeman/foreman-installer/pull/407#issue-343739127

File: config/foreman.hiera/scenario/katello.yaml
Comment (ehelms) [inline config/foreman.hiera/scenario/katello.yaml]: Is this the correct way to handle this file? This worked when I tested it so I assume yes :)

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [reply katello/hooks/pre/30-upgrade.rb:101]: Why is that needed? Regardless, it needs the same if conditions (and `if katello` isn't needed anymore then). Otherwise you break remote database support.

File: config/foreman.hiera/scenario/katello.yaml
Comment (ekohl) [reply config/foreman.hiera/scenario/katello.yaml]: In itself it's correct. I am slightly tempted to make it a bit more complex and go for `scenario/%{facts.kafo.scenario.id}/family/%{facts.os.family}-%{facts.os.release.major}.yaml` so we don't attempt to do this on EL8. Maybe that's just premature optimization though.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: This also removes `postgresql-libs` which is a dependency of `tfm-rubygem-postgresql`, right? Probably also the `python3-psycopg2` package.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: This is in KB
```suggestion
      $stderr.puts "The postgres upgrade requires at least #{postgres_size} KB of storage."
```

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: I'm wondering if this should be a fatal error as well.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb:45]: It surprises me that `yum remove postgresql-server` doesn't clean this up.

Comment (ehelms) [reply katello/hooks/pre/30-upgrade.rb:45]: Cause it gets overwritten by the module:

```
[root@centos7-katello-nightly-pg10 vagrant]# rpm -qlf /etc/systemd/system/postgresql.service 
file /etc/systemd/system/postgresql.service is not owned by any package
```

Comment (ekohl) [reply katello/hooks/pre/30-upgrade.rb:45]: Oh that's ugly. That should be using drop in files.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: Since https://github.com/theforeman/kafo/commit/64810c08a0fe08d5ef76b20144ce709ee52e6f9f I think this can just be `exit` (and below as well):

```suggestion
      exit 1
```

**Review by ekohl — COMMENTED** [review]
You should ensure `foreman-proxy-content.yaml` has the same content as `katello.yaml` once we add pulpcore. Can be a symlink. Currently we define it in the module but I'd like to also define it via hiera.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: Wondering whether this should reuse `stop_services`

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: Probably good to add a variable `has_local_postgresql` and reuse that.

**Comment by ehelms** [comment]
I went ahead and updated to do `scenario/<name>/family/<os-name>-<version>.yaml` since EL8 isn't that far off and this prevents having to re-do this part later.

There are improvements to the overall upgrade workflow we've mentioned but I would like to not block the upgrade effort on them. My preferred workflow would be:

 * merge this when it feels robust to get nightly onto new postgres and upgrade workflow
 * tackle other upgrade issues for Foreman 2.0
 * use fixing upgrade issues to handle forced postgres upgrade for Foreman scenario

The use of the `syspaths` RPMs should mean we don't have to touch foreman-maintain.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: I think this should get the disk space available on `/var/opt/rh/rh-postgresql10/lib/pgsql/data` (or the closest path that does exist). Total disk space isn't quite correct.
```ruby
new_postgres_dir = '/var/opt/rh/rh-postgresql10/lib/pgsql/data'
until File.directory?(new_postgres_dir)
  new_postgres_dir = File.dirname(new_postgres_dir)
end
available_space = `df -k --output=avail "#{new_postgres_dir}"`.split("\n").last.to_i
```

In `katello/hooks/pre_validations/30-mongo_storage_engine.rb` there's similar code so it may make sense to create a helper for this at some point :)

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb:46]: Since this is a pre hook, the installer runs after it. That ensures all `-syspaths` are installed so is this really needed? Does `rh-postgresql10-syspath` even exist as a package?

**Review by ekohl — COMMENTED** [review]
Overall this looks very good. I'd like to also upgrade vanilla Foreman before 2.0 hits, but I'd be happy to experiment with upgrading just katello first given the immediate need there.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [reply katello/hooks/pre/30-upgrade.rb]: To make it really EL-8 ready you can add `facts[:os][:release][:major] == '7'` as a condition.

File: katello/hooks/pre/30-upgrade.rb
Comment (ehelms) [reply katello/hooks/pre/30-upgrade.rb:46]: We want syspaths for the services and the client bits so I think using the overall package is ok.

```
rh-postgresql10-syspaths.x86_64 : System-wide wrappers for the rh-postgresql10
```

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: `--exclude-type=tmpfs` is technically not really effective because you ask it for a specific directory. Same for `--total`.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: Any reason you're not using my code? My version detected the separate partition `/var` for example where yours doesn't.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [reply katello/hooks/pre/30-upgrade.rb]: Alternative implementation that doesn't require shelling out:
```ruby
def available_bytes(directory)
  mountpoints = facts[:mountpoints]
  until (mountpoint = mountpoints[directory])
    directory = File.dirname(directory)
  end
  mountpoint[:available_bytes]
end
```

File: katello/hooks/pre/30-upgrade.rb
Comment (ehelms) [reply katello/hooks/pre/30-upgrade.rb]: I was not clear from looking at your code `how` it detected the separate partition without erroring when it did not exist _and_ then checked the base filesystem.

Comment (ekohl) [reply katello/hooks/pre/30-upgrade.rb]: It started with the directory and went up until the directory actually exists. You can give `df` any directory and it'll find the correct mountpoint as long as it exists. If you keep going up, you'll end up with `/` and that always exists:
```
[1] pry(main)> File.dirname('/doesnotexist')
=> "/"
```

Comment (ehelms) [reply katello/hooks/pre/30-upgrade.rb]: Not on my systems, which is part of why I didn't go that route:

```
irb(main):001:0> File.dirname('/var/foo/var/this/I/am/not/real')
=> "/var/foo/var/this/I/am/not"
```

Comment (ekohl) [reply katello/hooks/pre/30-upgrade.rb]: That's why there was the `until` loop :)
```
[1] pry(main)> directory = '/var/foo/var/this/I/am/not/real'
=> "/var/foo/var/this/I/am/not/real"
[2] pry(main)> until File.directory?(directory)
[2] pry(main)*   directory = File.dirname(directory)
[2] pry(main)* end  
=> nil
[3] pry(main)> puts directory
/var
=> nil
```

File: katello/hooks/pre/30-upgrade.rb
Comment (ehelms) [reply katello/hooks/pre/30-upgrade.rb]: Using this requires passing `available_bytes(:/)` if the directory does not exist? Just understanding how this works for the default case:

e.g
```
{:/=>{:available=>"34.04 GiB", :available_bytes=>36553474048, :capacity=>"13.26%", :device=>"/dev/vda1", :filesystem=>"ext4", :options=>["rw", "seclabel", "relatime", "data=ordered"], :size=>"39.25 GiB", :size_bytes=>42140401664, :used=>"5.20 GiB", :used_bytes=>5586927616}, :"/dev"=>{:available=>"3.81 GiB", :available_bytes=>4092948480, :capacity=>"0%", :device=>"devtmpfs", :filesystem=>"devtmpfs", :options=>["rw", "seclabel", "nosuid", "size=3997020k", "nr_inodes=999255", "mode=755"], :size=>"3.81 GiB", :size_bytes=>4092948480, :used=>"0 bytes", :used_bytes=>0}, :"/dev/hugepages"=>{:available=>"0 bytes", :available_bytes=>0, :capacity=>"100%", :device=>"hugetlbfs", :filesystem=>"hugetlbfs", :options=>["rw", "seclabel", "relatime"], :size=>"0 bytes", :size_bytes=>0, :used=>"0 bytes", :used_bytes=>0}, :"/dev/mqueue"=>{:available=>"0 bytes", :available_bytes=>0, :capacity=>"100%", :device=>"mqueue", :filesystem=>"mqueue", :options=>["rw", "seclabel", "relatime"], :size=>"0 bytes", :size_bytes=>0, :used=>"0 bytes", :used_bytes=>0}, :"/dev/pts"=>{:available=>"0 bytes", :available_bytes=>0, :capacity=>"100%", :device=>"devpts", :filesystem=>"devpts", :options=>["rw", "seclabel", "nosuid", "noexec", "relatime", "gid=5", "mode=620", "ptmxmode=000"], :size=>"0 bytes", :size_bytes=>0, :used=>"0 bytes", :used_bytes=>0}, :"/dev/shm"=>{:available=>"3.82 GiB", :available_bytes=>4100939776, :capacity=>"0.00%", :device=>"tmpfs", :filesystem=>"tmpfs", :options=>["rw", "seclabel", "nosuid", "nodev"], :size=>"3.82 GiB", :size_bytes=>4100960256, :used=>"20.00 KiB", :used_bytes=>20480}, :"/run"=>{:available=>"3.81 GiB", :available_bytes=>4091801600, :capacity=>"0.22%", :device=>"tmpfs", :filesystem=>"tmpfs", :options=>["rw", "seclabel", "nosuid", "nodev", "mode=755"], :size=>"3.82 GiB", :size_bytes=>4100960256, :used=>"8.73 MiB", :used_bytes=>9158656}, :"/run/user/1000"=>{:available=>"782.20 MiB", :available_bytes=>820195328, :capacity=>"0%", :device=>"tmpfs", :filesystem=>"tmpfs", :options=>["rw", "seclabel", "nosuid", "nodev", "relatime", "size=800972k", "mode=700", "uid=1000", "gid=1000"], :size=>"782.20 MiB", :size_bytes=>820195328, :used=>"0 bytes", :used_bytes=>0}, :"/sys/fs/cgroup"=>{:available=>"3.82 GiB", :available_bytes=>4100960256, :capacity=>"0%", :device=>"tmpfs", :filesystem=>"tmpfs", :options=>["ro", "seclabel", "nosuid", "nodev", "noexec", "mode=755"], :size=>"3.82 GiB", :size_bytes=>4100960256, :used=>"0 bytes", :used_bytes=>0}, :"/var/lib/nfs/rpc_pipefs"=>{:available=>"0 bytes", :available_bytes=>0, :capacity=>"100%", :device=>"sunrpc", :filesystem=>"rpc_pipefs", :options=>["rw", "relatime"], :size=>"0 bytes", :size_bytes=>0, :used=>"0 bytes", :used_bytes=>0}}
```

Comment (ekohl) [reply katello/hooks/pre/30-upgrade.rb]: It may need a `to_sym`. The idea is that you go "up" until it actually is found. By doing so, you're guaranteed to find the correct mountpoint that `mkdir -p` would also use. A minimal reproducer with debugging:
```ruby
require 'json'

# Taken from lib/kafo/base_context.rb
def symbolize(data)
  case data
  when Hash
    Hash[data.map { |key, value| [key.to_sym, symbolize(value)] }]
  when Array
    data.map { |v| symbolize(v) }
  else
    data
  end
end

FACTS = symbolize(JSON.parse(`facter --json`))

# Actual code

def available_bytes(directory)
  mountpoints = FACTS[:mountpoints]
  puts "Attempting #{directory}"
  until (mountpoint = mountpoints[directory.to_sym])
    raise Exception, "Could not find #{directory}" if directory == '/'
    directory = File.dirname(directory)
    puts "Attempting #{directory}"
  end
  puts "Found #{directory}"
  mountpoint[:available_bytes]
end

puts "Mountpoints: #{FACTS[:mountpoints].keys.inspect}"
puts available_bytes('/tmp/does/not/exist')
```

Output:
```
$ ruby disk_space.rb 
Mountpoints: [:/, :"/boot", :"/boot/efi", :"/dev", :"/dev/hugepages", :"/dev/mqueue", :"/dev/pts", :"/dev/shm", :"/run", :"/run/user/1000", :"/run/user/1000/gvfs", :"/tmp", :"/var/lib/nfs/rpc_pipefs"]
Attempting /tmp/does/not/exist
Attempting /tmp/does/not
Attempting /tmp/does
Attempting /tmp
Found /tmp
8328683520
```

Edit: the raise is redundant if everything works as it should, but prevents an infinite loop. I had this when I forgot the `.to_sym`.

File: katello/hooks/pre/30-upgrade.rb
Comment (ehelms) [inline katello/hooks/pre/30-upgrade.rb]: I tackled the default and `to_sym` this way -- seemed to work for me testing locally.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: The `.to_sym` needs to be here because `directory` is modified in the loop. Otherwise it won't find mountpoints like `/var`.

Comment (ehelms) [reply katello/hooks/pre/30-upgrade.rb]: Right, good catch

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: The until loop is also in `available_space` so doesn't need to be duplicated here.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: available space is in bytes and this in KB.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: I think we should move the disk space check to `katello/hooks/pre_validations` so we don't even start an upgrade unless there's sufficient disk space. Failing an upgrade leaves a feeling of an inconsistent state if it's not really.

**Review by ekohl — COMMENTED** [review]
I feel like kafo should provide a proper mechanism to provide generic helpers. Things like `has_local_postgresql?` and `available_space` are useful in various places. Going to think about that.

File: katello/hooks/pre/30-upgrade.rb
Comment (ehelms) [reply katello/hooks/pre/30-upgrade.rb]: Editing in two places for testing, I caught htis on my live system just not in git!

Comment (ekohl) [reply katello/hooks/pre/30-upgrade.rb]: I often have the same issue :( Technically it's nowadays possible to run straight from git but I never really do this.

**Comment by ehelms** [comment]
I moved to pre_validation which caused a bit of a re-factor of how I handled a few common methods. I'm trying to re-factor as little as possible with this change but I have seen some places improvements could be made.

File: katello/hooks/pre/30-upgrade.rb
Comment (ekohl) [inline katello/hooks/pre/30-upgrade.rb]: I think this should be a guard around `upgrade_step :postgresql_10_upgrade`. The difference is that otherwise on EL8 you'll have a file that says it did the postgresql 10 upgrade which it didn't really.

File: katello/hooks/boot/01-helpers.rb
Comment (ekohl) [inline katello/hooks/boot/01-helpers.rb:19]: If you move this to `HookContextExtension` you don't need to prefix it in the actual hooks.

File: katello/hooks/boot/01-helpers.rb
Comment (ekohl) [inline katello/hooks/boot/01-helpers.rb:57]: Can you move the `FIXME` from line 55 to just above this function? Otherwise someone might remove `local_postgresql?` as well. The `unless Kafo::HookContext.method_defined?(:param_value)` at the end of the file is also a risk.

Perhaps it's cleaner to add a separate module for things that shouldn't be upstreamed to Kafo itself.

**Review by ekohl — COMMENTED** [review]
Yes, the mongo hook can benefit a lot from these extensions.

File: katello/hooks/boot/01-helpers.rb
Comment (ehelms) [reply katello/hooks/boot/01-helpers.rb:19]: The function it calls, log_and_say is already in the helper so I wanted to co-locate them for the short term. Based on https://github.com/theforeman/foreman-installer/pull/407#discussion_r349763345 and my own observations, dropping the Helper for functions that are added via an extension as a separate re-factor would be a good clean up. And to provide some parity with Foreman hooks.

I would love to tackle that after this. Trying not to overload this particular change.

Comment (ekohl) [reply katello/hooks/boot/01-helpers.rb:19]: Fair enough. You can consider adding them in both places but this is already getting large.

**Comment by ehelms** [comment]
This is ready for another look.

File: katello/hooks/boot/01-helpers.rb
Comment (ekohl) [inline katello/hooks/boot/01-helpers.rb]: Perhaps this `unless` part should be around `def param_value` so we unconditionally include the `HookContextExtension`. Or just straight up remove the conditional part.

File: katello/hooks/boot/01-helpers.rb
Comment (ekohl) [inline katello/hooks/boot/01-helpers.rb:63]: How about adding the `fail_and_exit` this way:

```suggestion
  end
  
  def fail_and_exit(message)
    Kafo::Helpers.fail_and_exit(message)
  end
```

I'm also tempted to extract the helpers cleanup to a separate commit.

**Review by ekohl — COMMENTED** [review]
My biggest concern is the `unless` part, other than that I think this is pretty much ready to merge.

Off topic: I also want to take the `HookContextExtension` further. It was introduced by Martin but should probably be a more official way to customize the hook context. Nuking Kafo::Helpers and doing it right is a good way to merge foreman and katello hooks.

File: katello/hooks/boot/01-helpers.rb
Comment (ehelms) [reply katello/hooks/boot/01-helpers.rb]: Agreed, I dropped this.

File: katello/hooks/boot/01-helpers.rb
Comment (ehelms) [reply katello/hooks/boot/01-helpers.rb:63]: I think the behavior here preserves the way it was working the best and co-locates the log methods. In general, this will need clean up but I really want to do it *after* this postgres work is merged. I can file a Redmine issue to track it if that helps.

Comment (ekohl) [reply katello/hooks/boot/01-helpers.rb:63]: Probably a personal preference, but I'd have this PR as 2 commits: one that builds the infrastructure with `fail_and_exit` and the `HookContextExtension` and then one that only does the actual migration. Not a real difference in the end. Just a bit of personal preference in how I develop :)

**Review by ekohl — APPROVED** [review]
I didn't have a chance to test this yet, but it looks good.

**Pull request merged**
